### PR TITLE
types and conversions for porter app def

### PR DESCRIPTION
**Pull request by ianedwards**
## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Other (please describe):

## Pull request checklist

Please check if your PR fulfills the following requirements:

- [ ] If it's a backend change, tests for the changes have been added and `go test ./...` runs successfully from the root folder.
- [ ] If it's a frontend change, Prettier has been run
- [ ] Docs have been reviewed and added / updated if needed

### CODE REVIEW

File: dashboard/src/lib/porter-apps/services.ts [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+  autoscalingValidator,
+  healthcheckValidator,
+  deserializeAutoscaling,
+  deserializeHealthCheck,
+  serializeAutoscaling,
+  serializeHealth,
+} from "./values";
+import { Service, ServiceType } from "@porter-dev/api-contracts";
+
+export const serviceStringValidator = z.object({
```
Comment (d-g-town) [inline dashboard/src/lib/porter-apps/services.ts:15]: worth adding comments to all these functions saying something like "this object allows us to block users from changing service values that are present in the porter.yaml"

File: dashboard/src/lib/porter-apps/services.ts [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+          cron: config.cron.value,
+        },
+      })
+    )
+    .exhaustive();
+}
+
+export function deserializeService(
+  service: SerializedService,
+  existing?: SerializedService
```
Comment (d-g-town) [inline dashboard/src/lib/porter-apps/services.ts:174]: maybe rename existing to override? If I'm understanding correctly, this implements overrides for the Porter yaml? So deserialize service would be called with `deserializeService(frontEndChanges, porterYaml)`?

File: dashboard/src/lib/porter-apps/index.ts [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+export const porterAppFormValidator = z.object({
+  app: porterAppValidator,
+  source: sourceValidator,
+});
+export type PorterAppFormData = z.infer<typeof porterAppFormValidator>;
+
+export function porterClientAppFromProto(
+  proto: PorterApp,
+  buildpacks?: Buildpack[]
```
[the remark this answers is not in the document]
Comment (ianedwards) [reply dashboard/src/lib/porter-apps/index.ts:61]: I might be able to remove this, but for now the problem is that the proto will only give the buildpack names, but I want to be able to use the full buildpack object we get on detection. Don't really want to have to make the detectBuildpacks API call in this function.

If we don't need all the information that detectBuildpacks gives us when calling apply, then I'll remove it.

**Comment by d-g-town** [comment]
Generally looks good, but would be really helpful to start commenting the front-end code like we do the backend. Maybe to start, could you add comments to all the fleshier methods explaining what they are used for?  And distinguishing between a serialized vs service service? 

Would be very helpful for someone like me who has less familiarity with typescript to get a handle on what the logic should be accomplishing.

**Pull request merged**